controllers/machinehealthcheck_controller.go

```diff
@@ -2,61 +2,505 @@ package controllers
+			builder.WithPredicates(
+				predicate.Funcs{
+					// check for modified conditions on updates in order to prevent unneeded reconciliations
+					UpdateFunc: func(ev event.UpdateEvent) bool { return nodeUpdateNeedsReconcile(ev) },
+					// MHC reconciler is interested in deleted nodes... not sure why TBH?
+					DeleteFunc: func(_ event.DeleteEvent) bool { return true },
+					// create (new nodes don't have correct conditions yet), and generic events are not interesting for now
+					CreateFunc:  func(_ event.CreateEvent) bool { return false },
+					GenericFunc: func(_ event.GenericEvent) bool { return false },
+				},
+			),
+		)
+		bldr = bldr.Watches(
+			&source.Kind{Type: &v1beta1.Machine{}},
+			handler.EnqueueRequestsFromMapFunc(utils.MHCByMachineMapperFunc(mgr.GetClient(), mgr.GetLogger())),
+		)
+	}
+
+	return bldr.Complete(r)
+}
+
+func indexMachineByNodeName(object client.Object) []string {
+	machine, ok := object.(*v1beta1.Machine)
+	if !ok {
+		msg := fmt.Sprintf("Expected a machine for indexing field, got: %T", object)
+		ctrl.Log.WithName("machine indexer").Info(msg)
+		return nil
+	}
+
+	if machine.Status.NodeRef != nil {
+		return []string{machine.Status.NodeRef.Name}
+	}
+
+	return nil
+}
+
+// +kubebuilder:rbac:groups=machine.openshift.io,resources=machines,verbs=get;list;watch
+// +kubebuilder:rbac:groups=machine.openshift.io,resources=machinehealthchecks,verbs=get;list;watch;patch;update
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-func (r *MachineHealthCheckReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	//log := r.Log.WithValues("MachineHealthCheck", req.NamespacedName)
+func (r *MachineHealthCheckReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, returnErr error) {
+	log := r.Log.WithValues("MachineHealthCheck name", req.Name)
+
+	// get mhc
+	mhc := &v1beta1.MachineHealthCheck{}
+	err := r.Get(ctx, req.NamespacedName, mhc)
+	if err != nil {
+		log.Error(err, "failed getting Machine Health Check")
+		if apierrors.IsNotFound(err) {
+			return result, nil
+		}
+		return result, err
+	}
 
 	// update MHCChecker status
-	r.MHCChecker.UpdateStatus()
-	result := ctrl.Result{}
-
-	// fetch mhc
-	//mhc := &v1beta1.MachineHealthCheck{}
-	//err := r.Get(ctx, req.NamespacedName, mhc)
-	//if err != nil {
-	//	if apierrors.IsNotFound(err) {
-	//		log.Info("MachineHealthCheck not found", "name", req.Name, "namespace", req.Namespace)
-	//		return result, nil
-	//	}
-	//	log.Error(err, "failed fetching MachineHealthCheck", "name", req.Name, "namespace", req.Namespace)
-	//	return result, err
-	//}
-	//
-	//log.Info("reconciling MachineHealthCheck", "name", req.Name, "namespace", req.Namespace)
+	if err = r.MHCChecker.UpdateStatus(); err != nil {
+		return result, err
+	}
+
+	if !r.ReconcileMHC {
```

Comment:
shouldn't this be at the top of method ?

Reply:
no, the MHCChecker above isn't new

Reply:
What I mean is, in case the feature is disabled wouldn't it make sense to return immediately and skip other tests ?

Reply:
MHCChecker needs to always run 🤷🏼‍♂️